### ISSUE TITLE
compaction doesn't choose the newest supported format when compacting SSTables with different formats

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#3913]
*Installation details*
Scylla version (or git commit hash): 2.3.1

*Description*
Reported by a user who upgrades his cluster from 2.2.x to 2.3.0 and asked why he still sees `ka` SSTables with a creation timestamp long after the upgrade was over. More than that there were `la` SSTables with a timestamp older than than `ka` ones.

And reason for that was our compaction choosing the output SSTable format in obviously the wrong way:

```
beccasc1/scylla.log-2018.10.26-23:2018-10-26T22:06:23.605897+03:00 beccasc1 scylla: [shard 13] compaction - Compacting [/storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-11757-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-7597-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-9453-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-9197-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-8717-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-8269-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-10349-Data.db:level=0, ]
beccasc1/scylla.log-2018.10.26-23:2018-10-26T22:08:26.290890+03:00 beccasc1 scylla: [shard 13] compaction - Compacting [/storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-11085-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-8493-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-10573-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-9901-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-10125-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-10829-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-8941-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-11309-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-9677-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-11533-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-11661-Data.db:level=0, ]
beccasc1/scylla.log-2018.10.26-23:2018-10-26T22:09:05.139334+03:00 beccasc1 scylla: [shard 13] compaction - Compacting [/storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-5485-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-5229-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-4525-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-4813-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-4045-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-4301-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-3821-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-6733-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-6285-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-6509-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-6957-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-6029-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-7405-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-7181-Data.db:level=0, ]
beccasc1/scylla.log-2018.10.26-23:2018-10-26T22:09:40.544118+03:00 beccasc1 scylla: [shard 13] compaction - Compacted 7 sstables to [/storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-11853-Data.db:level=0, ]. 495060842 bytes to 143558176 (~28% of original) in 1821319ms = 0.08MB/s. ~3073280 total partitions merged to 871549.
beccasc1/scylla.log-2018.10.26-23:2018-10-26T22:19:38.160496+03:00 beccasc1 scylla: [shard 13] compaction - Compacted 32 sstables to [/storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-11821-Data.db:level=0, ]. 2005904627 bytes to 200073919 (~9% of original) in 2440249ms = 0.08MB/s. ~12918912 total partitions merged to 1234500.
```

**Comment by tgrabiec** [comment]
What's the output of `select peer, supported_features from system.peers;` on that node?

**Comment by gibsn** [comment]
Upgrade was from 2.2.x to 2.3.0

**Comment by vladzcloudius** [comment]
@gibsn My bad - I adjusted the description.

**Comment by tgrabiec** [comment]
@gibsn Did you see the problem before upgrading to 3.0-rc1, only after it, or you don't know?

**Comment by gibsn** [comment]
@tgrabiec before

**Comment by tgrabiec** [comment]
@gibsn Do you have a complete scylla log of that node, which includes at least the following points:
  1) the point in time when all nodes in the cluster were upgraded to 2.3
  2) any sstable compaction started after (1) which writes `ka` but was expected to write `la`

**Comment by gibsn** [comment]
@tgrabiec

1.  No, we upgraded to 2.3 on 09.21
2. I am not sure when compaction is expected to write `la`. Assuming that every compaction in 2.3 cluster is expected to write `la` the answer is yes.

**Comment by tgrabiec** [comment]
@gibsn The node should start to write la-only after all nodes are upgraded to 2.3. Do you currently see `ka` being created as well? If so, I'd recommend taking a core dump of that node and sending to us.

**Comment by gibsn** [comment]
@tgrabiec nope, only `la` are being created now. The `ka` tables are the ones left after a major compaction for the whole time window (we use TWCS) happened. And also the whole cluster is 3.0.rc1 now

**Comment by vladzcloudius** [comment]
@tgrabiec The problem is not that `ka` sstables were created due to memtable flushing after the upgrade - they were created as a result of compactions that were involving previously created `ka` sstables after the upgrade (see the opening message).

**Comment by tgrabiec** [comment]
@vladzcloudius In both cases the version is determined using `get_highest_supported_format()`, the fact that the compaction involved `ka` shouldn't matter.

**Comment by vladzcloudius** [comment]
@tgrabiec Sometimes the theory doesn't match the reality. We only know what we saw.

**Comment by tgrabiec** [comment]
Please send complete log (as large as possible) which includes the suspicious compactions.

**Comment by gibsn** [comment]
https://transfer.sh/I24EO/scylla.log-2018.11.02.tar.gz

**Comment by tgrabiec** [comment]
I checked the logs posted by @gibsn, and looked at the oldest record of `ka` sstable being written. It looks like it belonged to a long-running compaction, which was started soon after the node was booted, but before the node noticed that the cluster supports LA sstables. Here's the relevant subset of the log:

```
2018-11-02T14:49:56.533798+03:00 beccasc1 scylla: Scylla version 2.3.0-0.20180917.178f870a0 starting ...

2018-11-02T14:50:47.265717+03:00 beccasc1 scylla: [shard 13] compaction - Compacting [/storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/la-10317-big-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/la-13549-big-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/la-20973-big-Data.db:level=0, /storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/la-17645-big-Data.db:level=0, ]

2018-11-02T14:50:57.548443+03:00 beccasc1 scylla: [shard 0] gossip - Feature LA_SSTABLE_FORMAT is enabled

...

2018-11-02T23:29:14.701700+03:00 beccasc1 scylla: [shard 13] compaction - Compacted 4 sstables to [/storage/scylla/becca/events-a4d6c730d3c411e894da00000000001c/becca-events-ka-22445-Data.db:level=0, ]. 13491355475 bytes to 13016747927 (~96% of original) in 31107438ms = 0.40MB/s. ~12399232 total partitions merged to 7567746.
```

The compaction ran for hours. 

It is expected that it wrote 'ka', since the format is chosen when it starts.

**Comment by vladzcloudius** [comment]
@tgrabiec Two questions: 
1) How do you know that these two are related? 
2) Assuming they are indeed related - how come the compaction of 4 `la` sstables resulted in the `ka` sstable? Where these `la` came from if, according to you, when the compaction started the node still "thought" it only supports `ka` and in that case I'd expect memtable flushes generating `ka` sstables. 

Could you clarify that, please?

**Comment by tgrabiec** [comment]
Ad 1)

I deduce that from sstable generation numbers. 

We see the following sequence of messages on shard 13:

```
T0) Compacting ...
T1) gossip - Feature LA_SSTABLE_FORMAT is enabled
T2) Compacting ...
T3) Compacted ... into 22541
T4) Compacted ... into 22445
```

Because T4's generation is smaller than T3's, it must have started before T3, and there is only one such compaction, the one which started before T1.

Ad 2)

When the node boots, it assumes nothing about the cluster, until gossip determines the status of all the nodes.

The problem is that we allow compaction to start too early, before we determine cluster features on startup.

**Comment by avikivity** [comment]
The problem is deeper, we need to read sstables to determine gossip features but we need gossip features to know which sstable version we ought to write.

We can fix with a hack: sstable_version = max ( gossip sstable version, sstable versions on disk )

We should also hardcode it to mc in master so the problem is eliminated for future versions.

**Comment by tgrabiec** [comment]
On Tue, Jan 1, 2019 at 5:10 PM Avi Kivity <notifications@github.com> wrote:

> The problem is deeper, we need to read sstables to determine gossip
> features but we need gossip features to know which sstable version we ought
> to write.
>
So, we shouldn't start to write until we determine gossip features.

For a bootstrapping node that should already be the case, the node joins
the ring after it gossips.

For a restarting node, we should fix this bug, and start compaction only
after we read our state from system tables.

We can fix with a hack: sstable_version = max ( gossip sstable version,

**Comment by bhalevy** [comment]
@tgrabiec / @haaawk I don't think issue can happen anymore with #4205 fixed.  Please confirm, and if so we can close this issue.

**Comment by haaawk** [comment]
@bhalevy I think it should be fixed but someone has to test that. Do we have a reproducer?

**Comment by bhalevy** [comment]
@haaawk regarding a reproducer, I don't think we have an explicit test for that.  IIRC, we saw compaction generating sstables with la/ka format simply when restarting a node in a cluster that supports mc format.

**Comment by tgrabiec** [comment]
Fixed in 5dc3f5ea33f2c5a0a1a4c69e5a7a1fb5dc10c113
